rr_common/src/nn_planner/nn_planner.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python
+
+import rospy
+from cv_bridge import CvBridge, CvBridgeError
+from keras.models import load_model
+from sensor_msgs.msg import Image
+from rr_platform.msg import steering as Steering
+from rr_platform.msg import speed as Speed
+import cv2
+import numpy as np
+import os, sys
+from tensorflow import get_default_graph
+
+path_to_model = ''
+model = None
+input_shape = None
+
+# print "********", input_shape
+tf_graph = get_default_graph()
+
+steering_angle = 0
+has_rect_message = False
+
+def plan(image_message):
+    bridge = CvBridge()
+    try:
+        cv_image = bridge.imgmsg_to_cv2(image_message)
+    except CvBridgeError as e:
+        print e
+
+    # use bottom half only
+    halfHeight = cv_image.shape[0] // 2
+    cv_image = cv_image[halfHeight:, :]
+
+    X = np.zeros((1,) + input_shape)
+    X[0] = cv2.resize(cv_image, (input_shape[1], input_shape[0]))
+
+    with tf_graph.as_default():
+        Y = model.predict(X, batch_size=1)
+        print '  '.join('%.2f' % y for y in Y.flat)
+
+    i = np.argmax(Y)
+    global steering_angle
+    steering_angle = [-0.3, -0.1, 0, 0.1, 0.3][i]
+    # steering_angle *= -1
+
+def plan_rect(image_message):
+    has_rect_message = True
+    plan(image_message)
+
+def plan_raw(image_message):
+    if not has_rect_message:
+        plan(image_message)
+
+
+if __name__ == "__main__":
+    rospy.init_node('nn_planner')
+
+    path_to_model = os.path.join(os.path.dirname(__file__), rospy.get_param('~model_path'))
+    model = load_model(path_to_model)
+    input_shape = model.get_layer(index=0).get_input_shape_at(0)[1:]
+
+
+    publisher = rospy.Publisher('/steering', Steering, queue_size=1)
+    speed_publisher = rospy.Publisher('/speed', Speed, queue_size=1)
+    rospy.Subscriber('/camera/image_rect', Image, plan_rect)
```

Comment:
You may not have a problem with this if the network runs fast enough, but only setting the queue size on image subscribers doesn't  always work quite how you'd expect. The default buffer size is too small for images and if the pipeline takes too long, there will be a significant lag (even though you'd expect it to only use the most recent image). See IGVC's subscriber as an example https://github.com/RoboJackets/igvc-software/blob/master/igvc/src/cnn/cnn.py#L75